tool/tsh/common/db.go

```diff
@@ -908,66 +962,48 @@ type databaseInfo struct {
 	// database corresponds to the db route and may be nil, so use GetDatabase
 	// instead of accessing it directly.
 	database types.Database
+	isActive bool
 	mu       sync.Mutex
 }
 
 // GetDatabase returns the cached database or fetches it using the db route and
 // caches the result.
-func (d *databaseInfo) GetDatabase(cf *CLIConf, tc *client.TeleportClient) (types.Database, error) {
-	if d.ServiceName == "" && cf.DatabaseService == "" &&
-		len(tc.Labels) == 0 && tc.PredicateExpression == "" {
-		return nil, trace.BadParameter("specify a database service by name, --labels, or --query")
-	}
+func (d *databaseInfo) GetDatabase(ctx context.Context, tc *client.TeleportClient) (types.Database, error) {
 	d.mu.Lock()
 	defer d.mu.Unlock()
 	if d.database != nil {
 		return d.database, nil
 	}
 	// holding mutex across the api call to avoid multiple redundant api calls.
-	var databases types.Databases
-	var err error
-	name := d.ServiceName
-	if name != "" {
-		databases, err = listDatabasesByName(cf.Context, tc, name)
-	} else {
-		name = cf.DatabaseService
-		// search by prefix if the db name comes from cli flag instead of cert.
-		databases, err = listDatabasesByPrefix(cf.Context, tc, name)
-	}
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-	db, err := chooseOneDatabase(cf, name, databases)
+	database, err := getDatabase(ctx, tc, d.ServiceName)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-
-	d.database = db
+	d.database = database
 	return d.database, nil
 }
 
-// chooseOneDatabase is a helper func for GetDatabase that returns either the
-// only database in a list of databases or returns a database that matches the
-// nameOrPrefix exactly, otherwise an error.
-func chooseOneDatabase(cf *CLIConf, nameOrPrefix string, databases types.Databases) (types.Database, error) {
-	if len(databases) == 1 {
-		return databases[0], nil
-	}
-	// Check if nameOrPrefix matches any database exactly and, if so, choose
+// chooseOneDatabase is a helper func that returns either the only database in a
+// list of databases or returns a database that matches the selector name
+// exactly, otherwise an error.
+func chooseOneDatabase(cf *CLIConf, databases types.Databases) (types.Database, error) {
+	selectors := newDatabaseResourceSelectors(cf)
+	// Check if the name matches any database exactly and, if so, choose
 	// that database over any others.
 	for _, db := range databases {
-		if db.GetName() == nameOrPrefix {
+		if db.GetName() == selectors.name {
 			return db, nil
 		}
 	}
+	// look for a single database with a matching discovered name label.
+	if dbs, ok := findDatabasesByDiscoveredName(databases, selectors.name); ok {
+		databases = dbs
```

Comment:
For readability and safety can we create a internal variable instead of using input param ?

Reply:
could you clarify what you mean?

Reply:
it doesn't modify the slice in any way, if that's the concern